tests/test-grammar-parser.cpp

```diff
@@ -56,104 +161,71 @@ term  ::= [0-9]+)""";
         {LLAMA_GRETYPE_END, 0},
+        {LLAMA_GRETYPE_RULE_REF, 10},
+        {LLAMA_GRETYPE_END, 0},
         {LLAMA_GRETYPE_RULE_REF, 1},
-        {LLAMA_GRETYPE_RULE_REF, 4},
-        {LLAMA_GRETYPE_ALT, 0},
+        {LLAMA_GRETYPE_RULE_REF, 5},
+        {LLAMA_GRETYPE_END, 0},
         {LLAMA_GRETYPE_RULE_REF, 1},
+        {LLAMA_GRETYPE_RULE_REF, 5},
+        {LLAMA_GRETYPE_ALT, 0},
         {LLAMA_GRETYPE_END, 0},
         {LLAMA_GRETYPE_CHAR, 45},
         {LLAMA_GRETYPE_CHAR_ALT, 43},
         {LLAMA_GRETYPE_CHAR_ALT, 42},
         {LLAMA_GRETYPE_CHAR_ALT, 47},
         {LLAMA_GRETYPE_RULE_REF, 3},
         {LLAMA_GRETYPE_END, 0},
-        {LLAMA_GRETYPE_RULE_REF, 5},
+        {LLAMA_GRETYPE_RULE_REF, 8},
+        {LLAMA_GRETYPE_END, 0},
         {LLAMA_GRETYPE_RULE_REF, 6},
+        {LLAMA_GRETYPE_RULE_REF, 8},
         {LLAMA_GRETYPE_ALT, 0},
         {LLAMA_GRETYPE_END, 0},
         {LLAMA_GRETYPE_CHAR, 48},
         {LLAMA_GRETYPE_CHAR_RNG_UPPER, 57},
-        {LLAMA_GRETYPE_RULE_REF, 7},
+        {LLAMA_GRETYPE_END, 0},
+        {LLAMA_GRETYPE_RULE_REF, 9},
+        {LLAMA_GRETYPE_RULE_REF, 11},
+        {LLAMA_GRETYPE_END, 0},
+        {LLAMA_GRETYPE_RULE_REF, 9},
+        {LLAMA_GRETYPE_RULE_REF, 11},
         {LLAMA_GRETYPE_ALT, 0},
-        {LLAMA_GRETYPE_CHAR, 48},
-        {LLAMA_GRETYPE_CHAR_RNG_UPPER, 57},
         {LLAMA_GRETYPE_END, 0},
-    };
+    });
 
-    index = 0;
-    for (auto rule : parsed_grammar.rules)
-    {
-        // compare rule to expected rule
-        for (uint32_t i = 0; i < rule.size(); i++)
-        {
-            llama_grammar_element element = rule[i];
-            llama_grammar_element expected_element = expected_rules[index];
-
-            // pretty print error message before asserting
-            if (expected_element.type != element.type || expected_element.value != element.value)
-            {
-                fprintf(stderr, "index: %u\n", index);
-                fprintf(stderr, "expected_element: %d, %u\n", expected_element.type, expected_element.value);
-                fprintf(stderr, "actual_element: %d, %u\n", element.type, element.value);
-                fprintf(stderr, "expected_element != actual_element\n");
-            }
-
-            assert(expected_element.type == element.type && expected_element.value == element.value);
-            index++;
-        }
-    }
-
-    const char *longer_grammar_bytes = R"""(
-    root  ::= (expr "=" ws term "\n")+
-    expr  ::= term ([-+*/] term)*
-    term  ::= ident | num | "(" ws expr ")" ws
-    ident ::= [a-z] [a-z0-9_]* ws
-    num   ::= [0-9]+ ws
-    ws    ::= [ \t\n]*
-    )""";
-
-    parsed_grammar = grammar_parser::parse(longer_grammar_bytes);
-
-    expected = {
+    verify_parsing(R"""(
+        root  ::= (expr "=" ws term "\n")+
+        expr  ::= term ([-+*/] term)*
+        term  ::= ident | num | "(" ws expr ")" ws
+        ident ::= [a-z] [a-z0-9_]* ws
+        num   ::= [0-9]+ ws
+        ws    ::= [ \t\n]*
+    )""", {
```

Comment:
If these grammar strings stayed the same, why did the expected rules output change so drastically? I'm surprised to see so many changes in the rules reference tags and whatnot -- unless my diff engine simply isn't lining stuff up correctly?

Reply:
Not sure if you mean diff from master or within branch. There's a massive diff because of the changes to `+` (and also have fixed a bug about trivial repeated targets)

Its rewrite goes from `S+ --> S' ::= S S' | S` to:

```
S+ --> S'     ::= Scopy Sstar
       Scopy  ::= S
       Sstar  ::= Scopy Sstar |
```

*Except* if S was trivial at the first place (e.g. single ref)

Reply:
Argh, something is fishy, found a couple of bugs, will ping here when it's done.

Reply:
This is frying my brain. Next step is to use the gbnf-validator in tests to check the rewrites have the intended behaviour (e.g. moving it to common and importing it from this test and the json schema converter test), but that's a "next week" item for me :-D

Reply:
> This is frying my brain. Next step is to use the gbnf-validator in tests to check the rewrites have the intended behaviour (e.g. moving it to common and importing it from this test and the json schema converter test), but that's a "next week" item for me :-D

Alternatively, an easier next step might be to add some end-to-end tests in `test-grammar-integration.cpp` to test expected pass-and-fail strings against grammars that use these quantifiers -- that's where I would want to go next on this. If you're currently feeling fried, I'm happy to write those tests and make a PR to your PR. :D

Reply:
@ochafik Opened up #6644 to explain more of what I was imagining with these integration tests.  I think you're right -- we'll still want to integrate all of this with the gbnf-validator, but this felt like a lower friction path in the moment. Maybe that's just because I've been doing more with the integration tests lately.

Reply:
Definitely welcoming help on this branch (I might be able to review things a bit tomorrow). I’d also like to test this on more complex grammars (e.g one of the ones listed on the grammar performance issue) to check there’s no real world regressions (checking large outputs with same seed)